servicemanager/servicemanager.go

```diff
@@ -264,34 +371,129 @@ func (sm *ServiceManager) doApplyService(service ServiceInfo) (err error) {
 			continue
 		}
 
-		if err := os.RemoveAll(oldService.ImagePath); err != nil {
+		_, _, err := sm.removeService(oldService)
+		if err != nil {
 			log.Errorf("Can't remove old service: %s", err)
+			continue
+		}
+	}
+
+	if err = sm.serviceInfoProvider.ActivateService(service.ServiceID, service.AosVersion); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (sm *ServiceManager) doRevertService(service ServiceInfo) error {
+	_, _, err := sm.removeService(service)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (sm *ServiceManager) removePendingServices(
+	extraServiceSpace int64, extraLayerSpace int64,
+) (deallocatedServiceSize, deallocatedLayerSize int64, err error) {
+	for deallocatedServiceSize < extraServiceSpace || deallocatedLayerSize < extraLayerSpace {
+		if len(sm.pendingRemove) == 0 {
+			return 0, 0, ErrNotEnoughMemory
+		}
+
+		freedServiceSize, freedLayersSize, err := sm.removeService(sm.pendingRemove[0])
+		if err != nil {
+			return 0, 0, aoserrors.Wrap(err)
 		}
 
-		if err := sm.serviceInfoProvider.RemoveService(oldService); err != nil {
-			log.Errorf("Can't remove old service from storage: %s", err)
+		deallocatedServiceSize += freedServiceSize
+
+		deallocatedLayerSize += freedLayersSize
+
+		if sm.isLayersSamePartition {
```

Comment:
You could do it once before return. But better to move it out of this function: to AllocateLayersSpace and tryToAllocateSpace.

Reply:
In this case, it would be better to leave it because if we have both layers and services on the same partition, we will exit the loop earlier.